### PR TITLE
Exposing logger and scope to generated and custom client

### PR DESCRIPTION
**Pull request by ravirajj**
- Creating TChannelClient struct to expose loggers/scopes with appropriate fields/tags.
- Replacing TChannelClient interface with TChannelCaller interface.

### CODE REVIEW

File: examples/example-gateway/clients/quux/quux.go [hunk 1]
```diff
@@ -21,16 +21,49 @@
...
+	"go.uber.org/zap"
+	"golang.org/x/net/context"
 )
 
 // Client is a custom client that does nothing yet
 type Client interface{}
 
-type quux struct{}
+type quux struct {
+	client *zanzibar.TChannelClient
```
Comment (ChuntaoLu) [inline examples/example-gateway/clients/quux/quux.go]: This client is meant to be a demo for a custom client. It doesn't make sense to fill in a tchannel client which should be defined in thrift.

File: codegen/templates/tchannel_client.tmpl [hunk 2]
```diff
@@ -161,6 +159,9 @@ type {{$clientName}} struct {
 			return respHeaders, err
 		{{else -}}
 			resp, err = {{.GenCodePkgName}}.{{title $svc.Name}}_{{title .Name}}_Helper.UnwrapResponse(&result)
+			if err != nil {
```
Comment (Raynos) [inline codegen/templates/tchannel_client.tmpl]: not all errors are `.Error()` ; default to `.Warn()` here instead.

File: codegen/templates/tchannel_client.tmpl [hunk 3]
```diff
@@ -150,6 +147,7 @@ type {{$clientName}} struct {
 			}
 		}
 		if err != nil {
+			logger.Warn("{{$clientName}}.{{$methodName}} ({{$serviceMethod}}) returned error", zap.Error(err))
```
Comment (Raynos) [inline codegen/templates/tchannel_client.tmpl]: we do not put meta data in the log message. all of these should be fields the log string needs to be static.

File: codegen/templates/tchannel_client.tmpl [hunk 4]
```diff
@@ -161,6 +159,9 @@ type {{$clientName}} struct {
 			return respHeaders, err
 		{{else -}}
 			resp, err = {{.GenCodePkgName}}.{{title $svc.Name}}_{{title .Name}}_Helper.UnwrapResponse(&result)
+			if err != nil {
+				logger.Error("Unable to unwrap {{$clientName}}.{{$methodName}} ({{$serviceMethod}}) response", zap.Error(err))
```
Comment (Raynos) [inline codegen/templates/tchannel_client.tmpl]: We do not want a dynamic log string here. Put them all in fields.

File: examples/example-gateway/clients/quux/quux.go [hunk 5]
```diff
@@ -21,16 +21,49 @@
...
+	zanzibar "github.com/uber/zanzibar/runtime"
+	"go.uber.org/zap"
+	"golang.org/x/net/context"
 )
 
 // Client is a custom client that does nothing yet
 type Client interface{}
 
-type quux struct{}
+type quux struct {
```
Comment (Raynos) [inline examples/example-gateway/clients/quux/quux.go]: I dont think this clients needs changing...

File: examples/example-gateway/endpoints/bar/endpoint-config.json [hunk 6]
```diff
@@ -20,7 +20,8 @@
 	},
 	"dependencies": {
 		"client": [
-			"bar"
+			"bar",
+			"quux"
```
Comment (Raynos) [inline examples/example-gateway/endpoints/bar/endpoint-config.json]: why this change ?

File: runtime/tchannel_client.go [hunk 7]
```diff
@@ -53,12 +55,13 @@ type tchannelClient struct {
 	timeout           time.Duration
 	timeoutPerAttempt time.Duration
 	routingKey        *string
-	loggers           map[string]*zap.Logger
+	Loggers           map[string]*zap.Logger
+	Scopes            map[string]tally.Scope
```
[the remark this answers is not in the document]
Comment (ravirajj) [reply runtime/tchannel_client.go]: Yes, I'm debating which approach to take.  Thought I'd put up this diff for feedback.

File: runtime/tchannel_client.go [hunk 8]
```diff
@@ -43,8 +43,10 @@ type TChannelClientOption struct {
 	RoutingKey        *string
 }
 
-// tchannelClient implements TChannelClient and makes outgoing Thrift calls.
-type tchannelClient struct {
+// TChannelClient implements TChannelCaller and makes outgoing Thrift calls.
+type TChannelClient struct {
+	TChannelCaller
```
[the remark this answers is not in the document]
Comment (ravirajj) [reply runtime/tchannel_client.go]: Trying to see if go made sure the struct implements the interface, but that may not be the case.

File: codegen/templates/tchannel_client.tmpl (hunk 3 above)
Comment (ravirajj) [reply codegen/templates/tchannel_client.tmpl]: Actually, they might be already there, since the logger scoped the method.

Comment (Raynos) [reply codegen/templates/tchannel_client.tmpl]: if you want to drop them then that's fine too. We still want a static string non-dynamic here.

File: runtime/tchannel_client.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (ravirajj) [reply runtime/tchannel_client.go]: @clu One problem with having TChannelHandler interface that exposes Logger and Scope methods is that it will need to expose the endpoint.EndpointID, endpoint.HandlerID and endpoint.Method too.  Similarly for TChannelClient too, it would need add ClientID, etc to the interface too.  Using struct makes it easier to expose these fields and the Logger and Scope field too.

File: codegen/templates/tchannel_client.tmpl [hunk 9]
```diff
@@ -8,12 +8,10 @@ import (
 	"strconv"
 	"time"
 
-	"go.uber.org/zap"	
+	"go.uber.org/zap"
 	"go.uber.org/zap/zapcore"
-
```
[the remark this answers is not in the document]
Comment (ravirajj) [reply codegen/templates/tchannel_client.tmpl]: Done.

File: codegen/templates/tchannel_client.tmpl [hunk 10]
```diff
@@ -8,12 +8,10 @@ import (
...
 	zanzibar "github.com/uber/zanzibar/runtime"
 	"github.com/uber/tchannel-go"
-
```
[the remark this answers is not in the document]
Comment (ravirajj) [reply codegen/templates/tchannel_client.tmpl]: Done.

File: codegen/templates/tchannel_client.tmpl (hunk 3 above)
Comment (ravirajj) [reply codegen/templates/tchannel_client.tmpl]: Done.

File: codegen/templates/tchannel_client.tmpl (hunk 2 above)
Comment (ravirajj) [reply codegen/templates/tchannel_client.tmpl]: Done.

File: codegen/templates/tchannel_client.tmpl (hunk 4 above)
Comment (ravirajj) [reply codegen/templates/tchannel_client.tmpl]: Done.

File: examples/example-gateway/clients/quux/quux.go (hunk 5 above)
Comment (ravirajj) [reply examples/example-gateway/clients/quux/quux.go]: Just using it as an example to demonstrate how to create a custom tchannel client, but I can remove it.

File: examples/example-gateway/endpoints/bar/endpoint-config.json (hunk 6 above)
Comment (ravirajj) [reply examples/example-gateway/endpoints/bar/endpoint-config.json]: Same. To demostrate creating a custom tchannel client.

File: runtime/tchannel_client.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (ChuntaoLu) [reply runtime/tchannel_client.go]: why `EndpointID`, `HandlerID` and `Method` are exposed on the client? `ClientID` seems reasonable, what else need to be exposed for the client?

File: examples/example-gateway/endpoints/bar/endpoint-config.json (hunk 6 above)
Comment (Raynos) [reply examples/example-gateway/endpoints/bar/endpoint-config.json]: understood, got it.

File: codegen/templates/tchannel_client.tmpl [hunk 11]
```diff
@@ -147,7 +148,7 @@ type {{$clientName}} struct {
 			}
 		}
 		if err != nil {
-			logger.Warn("{{$clientName}}.{{$methodName}} ({{$serviceMethod}}) returned error", zap.Error(err))
+			logger.Warn("Client call returned error", zap.Error(err))
```
Comment (Raynos) [inline codegen/templates/tchannel_client.tmpl]: nitpick: "TChannel Client call...`

File: runtime/tchannel_client.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (ravirajj) [reply runtime/tchannel_client.go]: To clarify:

The `TChannelEndpoint` would expose the `TChannelHandler`, `Logger()`, `Scope()`, `EndpointID`, `HandlerID` and `Method`.

The `TChannelClient` would expose the `TChannelCaller`, `Logger(method)`, `Scope(method)` and `ClientID`.

File: runtime/tchannel_client.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (ChuntaoLu) [reply runtime/tchannel_client.go]: You already implemented `Call` on `TChannelClient`, https://github.com/uber/zanzibar/pull/264/files/87083f25e7e5ada3e823cb8e13488306973590a7#diff-4eed9ca277749573af3f65a5ffd21480R120
